notion-distcopy/src/test/scala/com/ambiata/notion/distcopy/DistCopyUploadSpec.scala

```diff
@@ -65,19 +69,42 @@ Upload files from HDFS to S3
     z <- b.get.execute(s3Client)
   } yield r -> z ==== data -> data)
 
- def noSourceFile = propNoShrink((s3: S3Temporary, hdfs: HdfsTemporary) => (for {
-   c <- ConfigurationTemporary.random.conf
-   a <- s3.address.execute(s3Client)
-   p <- hdfs.path.run(c)
-   _ <- DistCopyJob.run(Mappings(Vector(UploadMapping(p, a))), distCopyConf(c, s3Client))
- } yield ()) must beFail)
-
- def targetExists = propNoShrink((s3: S3Temporary, hdfs: HdfsTemporary) => (for {
-   c <- ConfigurationTemporary.random.conf
-   a <- s3.address.execute(s3Client)
-   p <- hdfs.path.run(c)
-   _ <- Hdfs.write(p, "").run(c)
-   _ <- a.put("").execute(s3Client)
-   _ <- DistCopyJob.run(Mappings(Vector(UploadMapping(p, a))), distCopyConf(c, s3Client))
- } yield ()) must beFail)
+  def directory = prop((s3: S3Temporary, hdfs: HdfsTemporary, file1: FilePath, files1: List[FilePath], data: String) => for {
+    c  <- ConfigurationTemporary.random.conf
+    io =  LocationIO(c, s3Client)
+    a  <- s3.address.execute(s3Client)
+    to =  S3Location(a.bucket, a.key)
+    files = file1 +: files1
+    d  <- hdfs.path.run(c)
+    _  <- files.map(f => new Path(d, f.path)).traverse(Hdfs.write(_, data).run(c))
+    _  <- DistCopy.uploadDirectory(HdfsLocation(d.toString), to, io)
+    ks <- io.listS3(to)
+    ds <- ks.traverse(io.readUtf8)
+  } yield ds ==== files.as(data))
```

Comment:
There are some tests for the mappings creation in `DistCopySpec` but they obviously need to be completed.